File: lib/voice-engine/diff-classifier.ts
// PR Sprint 7.16 — Heuristic diff classifier port.
//
// 1:1 translation of
// Helm SEO/helm-adaptive-voice-engine/diff_classifier.py.
// Detects 7 patterns in MVP:
//   1. Banned vocab (single words removed in edit)
//   2. Banned phrases (multi-word AI-isms removed)
//   3. Hook length delta
//   4. Emoji count delta
//   5. Hashtag count delta
//   6. CTA pattern shift (statement ↔ question)
//   7. Paragraph length drift
//
// Diffs that don't match any heuristic produce zero signals in
// MVP. The Python source has a stub _enqueue_for_llm_batch for
// Phase 1.5 (LLM batch classifier); we preserve the same stub as
// enqueueForLlmBatch so the call site is ready when we ship that
// worker.

import {
  newSignal,
  type ContentType,
  type Platform,
  type Signal,
} from './types';

// ============================================================
// Banned vocab targets — direct copy from the Python source.
// Known AI-isms that get edited out get higher confidence;
// less-common words get lower confidence so they need more
// samples before the override moves.
// ============================================================

const COMMON_AI_BUZZWORDS = new Set([
  'leverage',
  'harness',
  'unlock',
  'empower',
  'elevate',
  'streamline',
  'seamlessly',
  'effortlessly',
  'intuitively',
  'robust',
  'comprehensive',
  'holistic',
  'synergy',
  'navigate',
  'explore',
  'embark',
]);

const COMMON_AI_PHRASES: string[] = [
  'excited to share',
  'excited to announce',
  'thrilled to share',
  'thrilled to announce',
  'humbled to',
  'humbled by',
  'dive into',
  'delve into',
  'unpack',
  'uncover',
  'game-changer',
  'cutting-edge',
  'state-of-the-art',
  "in today's fast-paced world",
  'in the digital age',
  'at the end of the day',
  'at its core',
  'at the heart of',
  "it's worth noting that",
  "let's break it down",
  "let's unpack",
];

// ============================================================
// Tokenization helpers
//
// JS RegExp doesn't support Python's broad \w-with-unicode by
// default, but for the edits we're classifying (English-mostly)
// this is sufficient. Cross-language signal classification is a
// Phase 2 problem.
// ============================================================

const WORD_RE = /[A-Za-z0-9']+/g;
// Unicode property escapes catch emoji (BMP + supplementary).
// /u flag is required.
const EMOJI_RE = /\p{Extended_Pictographic}/gu;
const HASHTAG_RE = /#\w+/g;

function tokenizeWords(text: string): string[] {
  return (text.toLowerCase().match(WORD_RE) ?? []).slice();
}

function countEmojis(text: string): number {
  return (text.match(EMOJI_RE) ?? []).length;
}

function countHashtags(text: string): number {
  return (text.match(HASHTAG_RE) ?? []).length;
}

function splitParagraphs(text: string): string[] {
  return text
    .split(/\n\n+/)
    .map((p) => p.trim())
    .filter((p) => p.length > 0);
}

function firstLine(text: string): string {
  const trimmed = text.trim();
  if (!trimmed) return '';
  return trimmed.split('\n')[0];
}

function firstSentence(text: string): string {
  const trimmed = text.trim();
  if (!trimmed) return '';
  const m = trimmed.match(/[.!?](\s|$)/);
  if (!m) return trimmed;
  // Index after the punctuation + matched trailing space if any.
  const endIdx = (m.index ?? 0) + m[0].length;
  return trimmed.slice(0, endIdx).trim();
}

// ============================================================
// Heuristics
// ============================================================

export interface ClassifyOpts {
  original: string;
  edited: string;
  platform: Platform;
  contentType: ContentType;
  postId?: string | null;
}

export function detectHookLengthChange(
  o: ClassifyOpts,
): Signal | null {
  const origHook = firstSentence(firstLine(o.original));
  const editedHook = firstSentence(firstLine(o.edited));
  const origWords = tokenizeWords(origHook).length;
  const editedWords = tokenizeWords(editedHook).length;
  if (origWords === 0 || Math.abs(editedWords - origWords) < 2) return null;
  return newSignal({
    source: 'edit_diff',
    platform: o.platform,
    contentType: o.contentType,
    dimension: 'hook_length',
    valueDelta: {
      original_hook_words: origWords,
      edited_hook_words: editedWords,
      delta: editedWords - origWords,
    },
    confidence: 0.85,
  });
}

export function detectBannedVocabChanges(o: ClassifyOpts): Signal[] {
  const orig = new Set(tokenizeWords(o.original));
  const edited = new Set(tokenizeWords(o.edited));
  const removed: string[] = [];
  for (const w of orig) if (!edited.has(w)) removed.push(w);
  return removed.map((word) =>
    newSignal({
      source: 'edit_diff',
      platform: o.platform,
      contentType: o.contentType,
      dimension: 'banned_vocab',
      valueDelta: {
        removed_word: word,
        is_known_buzzword: COMMON_AI_BUZZWORDS.has(word),
      },
      confidence: COMMON_AI_BUZZWORDS.has(word) ? 0.95 : 0.55,
    }),
  );
}

export function detectBannedPhraseChanges(o: ClassifyOpts): Signal[] {
  const origLower = o.original.toLowerCase();
  const editedLower = o.edited.toLowerCase();
  const signals: Signal[] = [];
  for (const phrase of COMMON_AI_PHRASES) {
    if (origLower.includes(phrase) && !editedLower.includes(phrase)) {
      signals.push(
        newSignal({
          source: 'edit_diff',
          platform: o.platform,
          contentType: o.contentType,
          dimension: 'banned_vocab',
          valueDelta: {
            removed_phrase: phrase,
            is_known_buzzword: true,
          },
          confidence: 0.98,
        }),
      );
    }
  }
  return signals;
}

export function detectEmojiCountChange(o: ClassifyOpts): Signal | null {
  const orig = countEmojis(o.original);
  const edited = countEmojis(o.edited);
  if (orig === edited) return null;
  return newSignal({
    source: 'edit_diff',
    platform: o.platform,
    contentType: o.contentType,
    dimension: 'emoji_usage',
    valueDelta: {
      original_emoji_count: orig,
      edited_emoji_count: edited,
      delta: edited - orig,
    },
    confidence: 0.9,
  });
}

export function detectHashtagCountChange(o: ClassifyOpts): Signal | null {
  const orig = countHashtags(o.original);
  const edited = countHashtags(o.edited);
  if (orig === edited) return null;
  return newSignal({
    source: 'edit_diff',
    platform: o.platform,
    contentType: o.contentType,
    dimension: 'hashtag_strategy',
    valueDelta: {
      original_hashtag_count: orig,
      edited_hashtag_count: edited,
      delta: edited - orig,
    },
    confidence: 0.9,
  });
}

export function detectCtaPatternShift(o: ClassifyOpts): Signal | null {
  const origLines = o.original.trim().split('\n');
  const editedLines = o.edited.trim().split('\n');
  const origLast = origLines.length ? origLines[origLines.length - 1] : '';
  const editedLast = editedLines.length
    ? editedLines[editedLines.length - 1]
    : '';
  const origIsQuestion = origLast.trimEnd().endsWith('?');
  const editedIsQuestion = editedLast.trimEnd().endsWith('?');
  if (origIsQuestion === editedIsQuestion) return null;
  return newSignal({
    source: 'edit_diff',
    platform: o.platform,
    contentType: o.contentType,
    dimension: 'cta_style',
    valueDelta: {
      original_was_question: origIsQuestion,
      edited_is_question: editedIsQuestion,
      preferred_style: editedIsQuestion ? 'question' : 'statement',
    },
    confidence: 0.75,
  });
}

export function detectParagraphLengthChange(o: ClassifyOpts): Signal | null {
  const origParas = splitParagraphs(o.original);
  const editedParas = splitParagraphs(o.edited);
  if (origParas.length === 0 || editedParas.length === 0) return null;
  const origAvg =
    origParas.reduce((sum, p) => sum + tokenizeWords(p).length, 0) /
    origParas.length;
  const editedAvg =
    editedParas.reduce((sum, p) => sum + tokenizeWords(p).length, 0) /
    editedParas.length;
  if (Math.abs(editedAvg - origAvg) < 5) return null;
  return newSignal({
    source: 'edit_diff',
    platform: o.platform,
    contentType: o.contentType,
    dimension: 'paragraph_length',
    valueDelta: {
      original_avg_words: Math.round(origAvg * 10) / 10,
      edited_avg_words: Math.round(editedAvg * 10) / 10,
      delta: Math.round((editedAvg - origAvg) * 10) / 10,
    },
    confidence: 0.7,
  });
}

const SINGLE_SIGNAL_HEURISTICS: Array<(o: ClassifyOpts) => Signal | null> = [
  detectHookLengthChange,
  detectEmojiCountChange,
  detectHashtagCountChange,
  detectCtaPatternShift,
  detectParagraphLengthChange,
];

// ============================================================
// Main entry point
// ============================================================

export function classifyDiff(opts: ClassifyOpts): Signal[] {
  if (opts.original.trim() === opts.edited.trim()) return [];

  const signals: Signal[] = [];
  // Multi-signal heuristics first (banned vocab + phrases).
  signals.push(...detectBannedVocabChanges(opts));
  signals.push(...detectBannedPhraseChanges(opts));
  // Single-signal heuristics.
  for (const fn of SINGLE_SIGNAL_HEURISTICS) {
    const s = fn(opts);
    if (s) signals.push(s);
  }
  // Attach post_id when caller provided one.
  const withPostId: Signal[] = opts.postId
    ? signals.map((s) => ({ ...s, postId: opts.postId! }))
    : signals;
  // Stub: queue unclassified diffs for future LLM batch.
  if (withPostId.length === 0) {
    enqueueForLlmBatch(opts);
  }
  return withPostId;
}

// Stub for Phase 1.5 LLM batch classifier. No-op in MVP.
function enqueueForLlmBatch(_opts: ClassifyOpts): void {
  // In Phase 1.5: persist (original, edited, platform,
  // content_type) to a job queue. A nightly worker runs an LLM
  // classifier and emits supplemental signals into the feedback
  // loop. For MVP we drop the diff silently.
  return;
}

File: Helm SEO/helm-adaptive-voice-engine/diff_classifier.py
"""
diff_classifier.py
==================

Heuristic diff classifier for the Helm Adaptive Voice Engine v1.0.

Compares an originally-generated draft against the user's edited version,
extracts structured Signal objects per dimension. Heuristics only in MVP;
LLM batch classifier deferred to Phase 1.5.

Coverage target: heuristics handle ~70-80% of common edit patterns. Diffs
that don't match any heuristic produce no signals in MVP (the queue stub for
the future LLM batch is in place but the consumer is not built yet).

Version: 1.0 (MVP Phase 1, heuristics only)
"""

from __future__ import annotations

import re
from uuid import UUID

from client_context import (
    ContentType,
    Dimension,
    Platform,
    Signal,
    SignalSource,
)


# ============================================================================
# Banned vocab heuristic targets
# ============================================================================

# Words/phrases the system already knows tend to get edited out. These get
# higher confidence detection. Other removed words produce signals with lower
# confidence (so they don't move banned_vocab until many samples confirm).
COMMON_AI_BUZZWORDS: set[str] = {
    "leverage", "harness", "unlock", "empower", "elevate", "streamline",
    "seamlessly", "effortlessly", "intuitively", "robust", "comprehensive",
    "holistic", "synergy", "navigate", "explore", "embark",
    # Multi-word phrases handled separately below.
}

COMMON_AI_PHRASES: list[str] = [
    "excited to share", "excited to announce", "thrilled to share",
    "thrilled to announce", "humbled to", "humbled by",
    "dive into", "delve into", "unpack", "uncover",
    "game-changer", "cutting-edge", "state-of-the-art",
    "in today's fast-paced world", "in the digital age",
    "at the end of the day", "at its core", "at the heart of",
    "it's worth noting that", "let's break it down", "let's unpack",
]


# ============================================================================
# Tokenization helpers
# ============================================================================

WORD_RE = re.compile(r"\b[\w']+\b")
EMOJI_RE = re.compile(
    "["
    "\U0001F300-\U0001FAFF"
    "\U00002600-\U000027BF"
    "\U0001F1E0-\U0001F1FF"
    "]"
)
HASHTAG_RE = re.compile(r"#\w+")


def tokenize_words(text: str) -> list[str]:
    return WORD_RE.findall(text.lower())


def count_emojis(text: str) -> int:
    return len(EMOJI_RE.findall(text))


def count_hashtags(text: str) -> int:
    return len(HASHTAG_RE.findall(text))


def split_paragraphs(text: str) -> list[str]:
    return [p.strip() for p in text.split("\n\n") if p.strip()]


def first_line(text: str) -> str:
    return text.strip().split("\n")[0] if text.strip() else ""


def first_sentence(text: str) -> str:
    text = text.strip()
    if not text:
        return ""
    match = re.search(r"[.!?](\s|$)", text)
    return text[: match.end()].strip() if match else text


# ============================================================================
# Heuristics
# ============================================================================

def detect_hook_length_change(
    original: str, edited: str, platform: Platform, content_type: ContentType
) -> Signal | None:
    orig_hook = first_sentence(first_line(original))
    edited_hook = first_sentence(first_line(edited))

    orig_words = len(tokenize_words(orig_hook))
    edited_words = len(tokenize_words(edited_hook))

    if orig_words == 0 or abs(edited_words - orig_words) < 2:
        return None

    return Signal(
        source=SignalSource.EDIT_DIFF,
        platform=platform,
        content_type=content_type,
        dimension=Dimension.HOOK_LENGTH,
        value_delta={
            "original_hook_words": orig_words,
            "edited_hook_words": edited_words,
            "delta": edited_words - orig_words,
        },
        confidence=0.85,
    )


def detect_banned_vocab_changes(
    original: str, edited: str, platform: Platform, content_type: ContentType
) -> list[Signal]:
    """Detect single words removed in the edit. Multi-word phrases handled separately."""
    orig_words = set(tokenize_words(original))
    edited_words = set(tokenize_words(edited))
    removed = orig_words - edited_words

    signals: list[Signal] = []
    for word in removed:
        is_known_buzzword = word in COMMON_AI_BUZZWORDS
        signals.append(
            Signal(
                source=SignalSource.EDIT_DIFF,
                platform=platform,
                content_type=content_type,
                dimension=Dimension.BANNED_VOCAB,
                value_delta={
                    "removed_word": word,
                    "is_known_buzzword": is_known_buzzword,
                },
                confidence=0.95 if is_known_buzzword else 0.55,
            )
        )
    return signals


def detect_banned_phrase_changes(
    original: str, edited: str, platform: Platform, content_type: ContentType
) -> list[Signal]:
    """Detect known multi-word AI phrases removed in the edit."""
    orig_lower = original.lower()
    edited_lower = edited.lower()

    signals: list[Signal] = []
    for phrase in COMMON_AI_PHRASES:
        in_orig = phrase in orig_lower
        in_edited = phrase in edited_lower
        if in_orig and not in_edited:
            signals.append(
                Signal(
                    source=SignalSource.EDIT_DIFF,
                    platform=platform,
                    content_type=content_type,
                    dimension=Dimension.BANNED_VOCAB,
                    value_delta={
                        "removed_phrase": phrase,
                        "is_known_buzzword": True,
                    },
                    confidence=0.98,
                )
            )
    return signals


def detect_emoji_count_change(
    original: str, edited: str, platform: Platform, content_type: ContentType
) -> Signal | None:
    orig_count = count_emojis(original)
    edited_count = count_emojis(edited)

    if orig_count == edited_count:
        return None

    return Signal(
        source=SignalSource.EDIT_DIFF,
        platform=platform,
        content_type=content_type,
        dimension=Dimension.EMOJI_USAGE,
        value_delta={
            "original_emoji_count": orig_count,
            "edited_emoji_count": edited_count,
            "delta": edited_count - orig_count,
        },
        confidence=0.9,
    )


def detect_hashtag_count_change(
    original: str, edited: str, platform: Platform, content_type: ContentType
) -> Signal | None:
    orig_count = count_hashtags(original)
    edited_count = count_hashtags(edited)

    if orig_count == edited_count:
        return None

    return Signal(
        source=SignalSource.EDIT_DIFF,
        platform=platform,
        content_type=content_type,
        dimension=Dimension.HASHTAG_STRATEGY,
        value_delta={
            "original_hashtag_count": orig_count,
            "edited_hashtag_count": edited_count,
            "delta": edited_count - orig_count,
        },
        confidence=0.9,
    )


def detect_cta_pattern_shift(
    original: str, edited: str, platform: Platform, content_type: ContentType
) -> Signal | None:
    """Detect if the CTA shifted from statement to question or vice versa."""
    orig_last = original.strip().split("\n")[-1] if original.strip() else ""
    edited_last = edited.strip().split("\n")[-1] if edited.strip() else ""

    orig_is_question = orig_last.rstrip().endswith("?")
    edited_is_question = edited_last.rstrip().endswith("?")

    if orig_is_question == edited_is_question:
        return None

    return Signal(
        source=SignalSource.EDIT_DIFF,
        platform=platform,
        content_type=content_type,
        dimension=Dimension.CTA_STYLE,
        value_delta={
            "original_was_question": orig_is_question,
            "edited_is_question": edited_is_question,
            "preferred_style": "question" if edited_is_question else "statement",
        },
        confidence=0.75,
    )


def detect_paragraph_length_change(
    original: str, edited: str, platform: Platform, content_type: ContentType
) -> Signal | None:
    orig_paragraphs = split_paragraphs(original)
    edited_paragraphs = split_paragraphs(edited)

    if not orig_paragraphs or not edited_paragraphs:
        return None

    orig_avg = sum(len(tokenize_words(p)) for p in orig_paragraphs) / len(orig_paragraphs)
    edited_avg = sum(len(tokenize_words(p)) for p in edited_paragraphs) / len(edited_paragraphs)

    if abs(edited_avg - orig_avg) < 5:
        return None

    return Signal(
        source=SignalSource.EDIT_DIFF,
        platform=platform,
        content_type=content_type,
        dimension=Dimension.PARAGRAPH_LENGTH,
        value_delta={
            "original_avg_words": round(orig_avg, 1),
            "edited_avg_words": round(edited_avg, 1),
            "delta": round(edited_avg - orig_avg, 1),
        },
        confidence=0.7,
    )


# ============================================================================
# Main entry point
# ============================================================================

SINGLE_SIGNAL_HEURISTICS = [
    detect_hook_length_change,
    detect_emoji_count_change,
    detect_hashtag_count_change,
    detect_cta_pattern_shift,
    detect_paragraph_length_change,
]


def classify_diff(
    *,
    original: str,
    edited: str,
    platform: Platform,
    content_type: ContentType,
    post_id: UUID | None = None,
) -> list[Signal]:
    """Run all heuristics on a (original, edited) pair and return signals.

    Args:
        original:     the draft as generated by the model
        edited:       the draft as the user saved it
        platform:     target platform of the post
        content_type: content type of the post
        post_id:      optional, links signal to its source post for audit

    Returns:
        List of Signal objects, one per detected change pattern. Empty list if
        no heuristic matched (the diff will not produce learning signals in MVP).
    """
    if original.strip() == edited.strip():
        return []

    signals: list[Signal] = []

    # Multi-signal heuristics (one diff can produce many signals)
    signals.extend(detect_banned_vocab_changes(original, edited, platform, content_type))
    signals.extend(detect_banned_phrase_changes(original, edited, platform, content_type))

    # Single-signal heuristics
    for heuristic in SINGLE_SIGNAL_HEURISTICS:
        signal = heuristic(original, edited, platform, content_type)
        if signal:
            signals.append(signal)

    # Attach post_id if provided (Signal is frozen, so we copy)
    if post_id:
        signals = [s.model_copy(update={"post_id": post_id}) for s in signals]

    # Stub: queue unclassified diffs for future LLM batch
    if not signals:
        _enqueue_for_llm_batch(original, edited, platform, content_type)

    return signals


def _enqueue_for_llm_batch(
    original: str,
    edited: str,
    platform: Platform,
    content_type: ContentType,
) -> None:
    """Stub for Phase 1.5 LLM batch classifier.

    In MVP: no-op. Diffs without heuristic matches don't produce signals.
    In Phase 1.5: persist (original, edited, platform, content_type) to a
    job queue. A nightly worker runs an LLM classifier and emits supplemental
    signals into the feedback loop.
    """
    return None
